bip-0174.mediawiki

```diff
@@ -0,0 +1,395 @@
+is the separator here <tt>0x00</tt> instead of <tt>0xff</tt>?'''
+The separator here is used to distinguish between each chunk of data. A separator of 0x00 would mean that
+the unserializer can read it as a key length of 0, which would never occur with actual keys. It can thus
+be used as a separator and allow for easier unserializer implementation.</ref>. The format
+of a record is as follows:
+
+<pre>
+<key length>|<key>|<value length>|<value>
+</pre>
+
+{| class="wikitable" style="width: auto; text-align: center; font-size: smaller; table-layout: fixed;"
+!Name
+!Type
+!Description
+|-
+| Key Length
+| Compact Size Unsigned Integer
+| Specify how long the key is
+|-
+| Key
+| byte[]
+| The key itself
+|-
+| Value Length
+| Compact Size Unsigned Integer
+| Specify how long the value is
+|-
+| Value
+| byte[]
+| The Value itself
+|}
+
+The format of each key-value map is as follows:
+
+<pre>
+<key-value pair>|<key-value pair>|...|0x00
+</pre>
+
+{| class="wikitable" style="width: auto; text-align: center; font-size: smaller; table-layout: fixed;"
+!Field Size
+!Name
+!Type
+!Value
+!Description
+!Format Example
+|-
+| 1+
+| Key-value pairs
+| Array of key-value pairs
+| varies
+| The key-value pairs.
+|-
+| 1
+| separator
+| char
+| <tt>0x00</tt>
+| Must be <tt>0x00</tt>.
+|}
+
+The first byte of each key specifies the type of the key-value pair. Some types are
+for global fields and other fields are for each input. The only required type in a
+PSBT is the transaction type, as defined below. The currently defined global types are as follows:
+
+{| class="wikitable" style="width: auto; text-align: center; font-size: smaller;
+table-layout: fixed;"
+!Number
+!Name
+!Key Data
+!Value Data
+!Format Example
+|-
+| <tt>0x00</tt>
+| Transaction
+| None. The key must only contain the 1 byte type.
+| The transaction in network serialization. The scriptSigs and
+witnesses for each input must be empty unless the input is finalized. The transaction
+must be in the witness serialization format as defined in BIP 144. A PSBT must have
+a transaction, otherwise it is invalid.
+| <pre>
+    <key length(0x01)>|0x00|<tx length>|<transaction>
+  </pre>
+|-
+| <tt>0x01</tt>
+| Redeem Script<ref>'''Why are redeem scripts and witness scripts global''' Redeem
+  scripts and witness scripts are global data to avoid duplication. Instead of specifying
+  a redeems script and witness script multiple times in inputs that need those scripts,
+  they are specified once in the global data.</ref>
+| The hash160 of the redeem script
+| A redeem script that will be needed to sign a Pay-To-Script-Hash input.
+| <pre>
+    <key length>|0x01|<hash160>|<redeem script length>|<redeem script>
+  </pre>
+|-
+| <tt>0x02</tt>
+| Witness Script
+| The sha256 hash of the witness script
+| A witness script that will be needed to sign a Pay-To-Witness-Script-Hash input.
+| <pre>
+    <key length>|0x02|<sha256>|<witness script length>|<witness script>
```

Comment:
maybe `<key length>` should include the type prefix,  then you can simply define `<key length>` as a the length of the `<key>`, and interpreters interpret the first byte of a key as a type.
That way there is less back-and-forth over the definition,  and its consistent.
I suspect this is how your implementation would work internally anyway.

For example, as your definition:
```
<key length> | <key(0x02 | sha256(<witness script>)) | <value length> | <value(witness script)>
```

Hell,  then you could simplify the entire defintion to simply `key` and `value`, with a line at the top here saying `key` and `length` are both prefixed by their `length` as a varint (or whatever).

Reply:
If you want to move past this now,  I could submit this as a PR if you'd like.

Reply:
That is what I originally meant. The latest changes should make that clearer.